### PR TITLE
Enable CNI

### PR DESCRIPTION
**Pull request by HuKeping**
The test script would be:
```
#! /bin/bash

set -x
set -e

# it is unlikely we'd use the notary
unset DOCKER_CONTENT_TRUST

# contaienrID of course.
CID=`docker run -d ubuntu:14.04 sleep 10000`

PID=`docker inspect -f "{{.State.Pid}}" $CID`

DEVICE_NAME="eth1"

# Use the bridge Ip, "docker0" for example, as the default gateway for container
Bridge=docker0
#BridgeIP=`ifconfig $Bridge | grep inet | grep -v inet6 | awk '{print $2}' |tr -d "addr:"`
BridgeIP="172.17.0.1"


mkdir -p /var/run/netns

# /var/run/netns/xxx is what `ip` wanted.
ln -s /proc/$PID/ns/net /var/run/netns/$PID

# set up veth pair A and B, attach A to bridge-docker0.
ip link add A type veth peer name B
brctl addif $Bridge A
ip link set A up

# add B into container's network namespace and hack it.
ip link set B netns $PID
ip netns exec $PID ip link set dev B name $DEVICE_NAME
ip netns exec $PID ip link set $DEVICE_NAME address 12:34:56:78:9a:bb
ip netns exec $PID ip addr add 172.17.42.101/16 dev $DEVICE_NAME
ip netns exec $PID ip link set $DEVICE_NAME up
#ip netns exec $PID ip route add default via $BridgeIP

# just in case we want do some thing
sleep 30

ip netns exec $PID ip link delete $DEVICE_NAME
sleep 30

# teardown
docker stop $CID
docker rm $CID

find -L /var/run/netns -type l -delete
ip link delete A
```

### CODE REVIEW

File: supervisor/hyperpod.go [hunk 1]
```diff
@@ -153,15 +181,85 @@ func (hp *HyperPod) initPodNetwork(c *Container) error {
...
+				glog.Error(err)
+				continue
+			}
+
+			conf := pod.UserInterface{
+				Bridge: bridge,
+				Ip:     update.Addr.LinkAddress.String(),
+			}
+
+			err = hp.vm.AddNic(update.Addr.LinkIndex, "eth1", conf)
```
Comment (HuKeping) [inline supervisor/hyperpod.go]: It is for test here, will change in the future.

File: hypervisor/devicemap.go [hunk 2]
```diff
@@ -506,6 +524,19 @@ func (ctx *VmContext) removeInterface() {
 	}
 }
 
+func (ctx *VmContext) removeInterfaceByLinkIndex(index int) {
+	for idx, nic := range ctx.devices.networkMap {
+
+		glog.V(1).Infof("==============>idx:nic is %v:%v", idx, nic)
```
Comment (HuKeping) [inline hypervisor/devicemap.go]: This is used to see what we have in the networkMap, it's a debug print, will remove in the future.

File: hypervisor/devicemap.go [hunk 3]
```diff
@@ -489,6 +489,24 @@ func (ctx *VmContext) releaseNetwork() {
...
+func (ctx *VmContext) releaseNetworkByLinkIndex(index int) {
+	var maps []pod.UserContainerPort
+
+	for _, c := range ctx.userSpec.Containers {
+		for _, m := range c.Ports {
+			maps = append(maps, m)
+		}
+	}
+
+	for idx, nic := range ctx.devices.networkMap {
```
Comment (gao-feng) [inline hypervisor/devicemap.go]: Don't need cycle here,
nic, ok := ctx.devices.networkMap[index]

File: hypervisor/devicemap.go [hunk 4]
```diff
@@ -506,6 +524,19 @@ func (ctx *VmContext) removeInterface() {
 	}
 }
 
+func (ctx *VmContext) removeInterfaceByLinkIndex(index int) {
+	for idx, nic := range ctx.devices.networkMap {
```
Comment (gao-feng) [inline hypervisor/devicemap.go]: same

File: hypervisor/devicemap.go (hunk 3 above)
Comment (HuKeping) [inline hypervisor/devicemap.go]: updated

**Comment by HuKeping** [comment]
Updated.

**Comment by gao-feng** [comment]
LGTM

**Pull request merged**